std/heap.zig

```diff
@@ -89,21 +78,56 @@ pub const DirectAllocator = struct {
 
                 return @intToPtr([*]u8, aligned_addr)[0..n];
             },
-            Os.windows => {
-                const amt = n + alignment + @sizeOf(usize);
-                const optional_heap_handle = @atomicLoad(?HeapHandle, &self.heap_handle, builtin.AtomicOrder.SeqCst);
-                const heap_handle = optional_heap_handle orelse blk: {
-                    const hh = os.windows.HeapCreate(0, amt, 0) orelse return error.OutOfMemory;
-                    const other_hh = @cmpxchgStrong(?HeapHandle, &self.heap_handle, null, hh, builtin.AtomicOrder.SeqCst, builtin.AtomicOrder.SeqCst) orelse break :blk hh;
-                    _ = os.windows.HeapDestroy(hh);
-                    break :blk other_hh.?; // can't be null because of the cmpxchg
-                };
-                const ptr = os.windows.HeapAlloc(heap_handle, 0, amt) orelse return error.OutOfMemory;
-                const root_addr = @ptrToInt(ptr);
-                const adjusted_addr = mem.alignForward(root_addr, alignment);
-                const record_addr = adjusted_addr + n;
-                @intToPtr(*align(1) usize, record_addr).* = root_addr;
-                return @intToPtr([*]u8, adjusted_addr)[0..n];
+            .windows => {
+                const w = os.windows;
+
+                // Although officially it's at least aligned to page boundary,
+                // Windows is known to reserve pages on a 64K boundary. It's
+                // even more likely that the requested alignment is <= 64K than
+                // 4K, so we're just allocating blindly and hoping for the best.
+                // see https://devblogs.microsoft.com/oldnewthing/?p=42223
+                const addr = w.VirtualAlloc(
+                    null,
+                    n,
+                    w.MEM_COMMIT | w.MEM_RESERVE,
+                    w.PAGE_READWRITE,
+                ) orelse return error.OutOfMemory;
+
+                // If the allocation is sufficiently aligned, use it.
+                if (@ptrToInt(addr) & (alignment - 1) == 0) {
+                    return @ptrCast([*]u8, addr)[0..n];
+                }
+
+                // If it wasn't, actually do an explicitely aligned allocation.
```

Comment:
This also needs to do a `VirtualFree` on `addr`, yeah?

Reply:
Yes, indeed.